packages/queued-request-controller/src/QueuedRequestMiddleware.ts

```diff
@@ -0,0 +1,178 @@
+import type {
+  AddApprovalRequest,
+  ApprovalRequest,
+} from '@metamask/approval-controller';
+import type { ControllerMessenger } from '@metamask/base-controller';
+import type { InfuraNetworkType } from '@metamask/controller-utils';
+import {
+  ApprovalType,
+  BUILT_IN_NETWORKS,
+  isNetworkType,
+} from '@metamask/controller-utils';
+import type { JsonRpcMiddleware } from '@metamask/json-rpc-engine';
+import { createAsyncMiddleware } from '@metamask/json-rpc-engine';
+import type {
+  NetworkClientId,
+  NetworkControllerFindNetworkClientIdByChainIdAction,
+  NetworkControllerGetNetworkClientByIdAction,
+  NetworkControllerGetStateAction,
+  NetworkControllerSetActiveNetworkAction,
+  NetworkControllerSetProviderTypeAction,
+} from '@metamask/network-controller';
+import { serializeError } from '@metamask/rpc-errors';
+import type { SelectedNetworkControllerSetNetworkClientIdForDomainAction } from '@metamask/selected-network-controller';
+import { SelectedNetworkControllerActionTypes } from '@metamask/selected-network-controller';
+import type { Json, JsonRpcParams, JsonRpcRequest } from '@metamask/utils';
+
+import type { QueuedRequestControllerEnqueueRequestAction } from './QueuedRequestController';
+import { QueuedRequestControllerActionTypes } from './QueuedRequestController';
+
+const isConfirmationMethod = (method: string) => {
+  const confirmationMethods = [
+    'eth_sendTransaction',
+    'wallet_watchAsset',
+    'wallet_switchEthereumChain',
+    'eth_signTypedData_v4',
+  ];
+
+  return confirmationMethods.includes(method);
+};
+
+/**
+ * Creates a JSON-RPC middleware for handling queued requests. This middleware
+ * intercepts JSON-RPC requests, checks if they require queueing, and manages
+ * their execution based on the specified options.
+ *
+ * @param options - Configuration options.
+ * @param options.messenger - A controller messenger used for communication with various controllers.
+ * @param options.useRequestQueue - A function that determines if the request queue feature is enabled.
+ * @returns The JSON-RPC middleware that manages queued requests.
+ */
+export const createQueuedRequestMiddleware = ({
+  messenger,
+  useRequestQueue,
+}: {
+  messenger: ControllerMessenger<
+    | QueuedRequestControllerEnqueueRequestAction
+    | NetworkControllerGetStateAction
+    | NetworkControllerSetActiveNetworkAction
+    | NetworkControllerSetProviderTypeAction
+    | NetworkControllerGetNetworkClientByIdAction
+    | NetworkControllerFindNetworkClientIdByChainIdAction
+    | SelectedNetworkControllerSetNetworkClientIdForDomainAction
+    | AddApprovalRequest,
+    never
+  >;
```

Comment:
Typing fix to follow pattern established in other controllers.

```suggestion
  messenger: QueuedRequestControllerMessenger;
```

Reply:
Required import statement changes:

```diff
@@ -1,8 +1,4 @@
-import type {
-  AddApprovalRequest,
-  ApprovalRequest,
-} from '@metamask/approval-controller';
-import type { ControllerMessenger } from '@metamask/base-controller';
+import type { ApprovalRequest } from '@metamask/approval-controller';
 import type { InfuraNetworkType } from '@metamask/controller-utils';
 import {
   ApprovalType,
@@ -11,20 +7,12 @@ import {
 } from '@metamask/controller-utils';
 import type { JsonRpcMiddleware } from '@metamask/json-rpc-engine';
 import { createAsyncMiddleware } from '@metamask/json-rpc-engine';
-import type {
-  NetworkClientId,
-  NetworkControllerFindNetworkClientIdByChainIdAction,
-  NetworkControllerGetNetworkClientByIdAction,
-  NetworkControllerGetStateAction,
-  NetworkControllerSetActiveNetworkAction,
-  NetworkControllerSetProviderTypeAction,
-} from '@metamask/network-controller';
+import type { NetworkClientId } from '@metamask/network-controller';
 import { serializeError } from '@metamask/rpc-errors';
-import type { SelectedNetworkControllerSetNetworkClientIdForDomainAction } from '@metamask/selected-network-controller';
 import { SelectedNetworkControllerActionTypes } from '@metamask/selected-network-controller';
 import type { Json, JsonRpcParams, JsonRpcRequest } from '@metamask/utils';
 
-import type { QueuedRequestControllerEnqueueRequestAction } from './QueuedRequestController';
+import type { QueuedRequestControllerMessenger } from './QueuedRequestController';
 import { QueuedRequestControllerActionTypes } from './QueuedRequestController';
```